### PR TITLE
Notify

### PR DESCRIPTION
**Pull request by tertsdiepraam**
This PR includes:
- The implementation of `ldns-notify` and `dnst notify`.
- Stelline in the env.
- Printing version and help from ldns argument parsing.

### CODE REVIEW

File: src/commands/notify.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        let mut config = dgram::Config::new();
+        config.set_max_retries(self.retries as u8);
+        let connection = dgram::Connection::with_config(UdpConnect::new(socket), config);
+        let connection = tsig::Connection::new(key, connection);
+
+        let req = RequestMessage::new(msg).unwrap();
+        let mut req = connection.send_request(req);
+    
+        let when = Local::now();
+        let msg = req.get_response().await.map_err(|e| format!("warning: could not send message to {socket}: {e}"))?;
```
Comment (mozzieongit) [inline src/commands/notify.rs:265]: I would rephrase this error, as I think it's misleading. The request was sent (at least in my test), but the error comes from validating the response from the server. So maybe more like: `warning: erroneous response from {socket}: {e}`. I came across this when using tsig, but the server didn't use tsig in the response.

Comment (mozzieongit) [reply src/commands/notify.rs:265]: Or will there be different errors? Some from failed sending and some from failed receiving?

**Review by mozzieongit — CHANGES_REQUESTED** [review]
It looked like you're mostly done, so I checked it out. There's a left-over TODO though.

File: src/commands/notify.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    #[arg(short = 'd', long = "debug")]
+    debug: bool,
+
+    /// Max number of retries
+    #[arg(short = 'r', long = "retries", default_value = "15")]
+    retries: usize,
+
+    // Hidden extra argument for `-?` to trigger help, which ldns supports.
+    #[arg(short = '?', action = ArgAction::Help, hide = true)]
+    compatible_help: (),
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/notify.rs:139]: Don't know where I got that from :smile:

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/notify.rs:139]: Ah this is where I got it from:

https://github.com/NLnetLabs/ldns/blob/b87b8e1d212b43f225891570ebc5045c6331dd8a/examples/ldns-notify.c#L282

But it's not part of the string here:

https://github.com/NLnetLabs/ldns/blob/b87b8e1d212b43f225891570ebc5045c6331dd8a/examples/ldns-notify.c#L193

So it's not supported indeed.

File: src/commands/notify.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // TODO: better error messages
+        let Some((mut name, rest)) = s.split_once(':') else {
+            return Err("invalid tsig string".into());
+        };
+
+        let mut key;
+        let mut algorithm;
+        if let Some((k, a)) = s.split_once(':') {
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/notify.rs:67]: Oof that's sloppy 😅

File: src/commands/mod.rs [hunk 4]
```diff
@@ -11,6 +12,14 @@ pub enum Command {
     #[command(name = "nsec3-hash")]
     Nsec3Hash(self::nsec3hash::Nsec3Hash),
 
+    /// Send a NOTIFY packet to DNS servers
+    ///
+    /// This tells them that an updated zone is available at the master servers. It can perform TSIG
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/mod.rs:17]: Is this just newer terminology or does it mean anything different? In the first case, I definitely agree. But this came from the manpage of ldns-notify, so decided to copy that first.

[the remark this answers is not in the document]
Comment (mozzieongit) [reply src/commands/mod.rs:17]: It's quite old terminology actually (which had been replaced by master/slave some decades ago), but is now changed to primary/seconday again, as in many other environments (git, parallel prog., etc.).
It means the same, and is known terminology, the ldns man pages are just old.

[the remark this answers is not in the document]
Comment (ximon18) [reply src/commands/mod.rs:17]: I think our preference is to avoid using words that are known to be offensive to some and use alternate replacement terms where available. Master and slave should indeed I believe be replaced by primary and secondary. [RFC 9499](https://datatracker.ietf.org/doc/html/rfc9499) states:

> Slave server: See "Secondary server".
> Master server: See "Primary server".

File: src/commands/notify.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/commands/notify.rs:265]: Both are possible yeah and unfortunately not clearly distinguished. I'll update the text to mention both possibilities.

File: src/commands/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/mod.rs:17]: Excellent! Then I'll do that, because that terminology should indeed disappear

File: src/commands/notify.rs [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        let connection = tsig::Connection::new(key, connection);
+
+        let req = RequestMessage::new(msg).unwrap();
+        let mut req = connection.send_request(req);
+
+        let when = Local::now();
+        let msg = req
+            .get_response()
+            .await
+            .map_err(|e| format!("warning: could not send message to {socket}: {e}"))?;
```
Comment (mozzieongit) [inline src/commands/notify.rs:290]: Here as well?
```suggestion
            .map_err(|e| format!("warning: reply was not received or erroneous from {socket}: {e}"))?;
```

File: src/commands/notify.rs [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    }
+}
+
+impl FromStr for TSigInfo {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // TODO: better error messages
+        let Some((mut name, rest)) = s.split_once(':') else {
+            return Err("invalid tsig string".into());
```
Comment (mozzieongit) [inline src/commands/notify.rs:61]: Maybe "invalid TSIG key format" or is the TODO just forgotten?

**Comment by mozzieongit** [comment]
FYI: ldns-notify supports `-v`

File: src/commands/notify.rs [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+    // #[arg(short = 'I', required = false)]
+    // source_address: (),
+    //
+    /// SOA version number to include
+    #[arg(short = 's')]
+    soa_version: Option<u32>,
+
+    /// A base64 tsig key and optional algorithm to include
+    #[arg(
+        short = 'y',
```
Comment (bal-e) [inline src/commands/notify.rs:124]: Why is this `-y`?  Do we want to change it to something more obvious, or should we keep it as-is for ldns compatibility?

File: src/commands/notify.rs [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+            for socket in hosts.port_iter(self.port) {
+                self.notify_host(&env, socket, msg.clone(), server, &tsig)
+                    .await;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn notify_host(
```
Comment (bal-e) [inline src/commands/notify.rs:351]: Could you document what this and the other internal functions do?

File: src/commands/notify.rs [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+
+    async fn notify_host_with_tsig(
+        &self,
+        env: impl Env,
+        socket: SocketAddr,
+        msg: Message<Vec<u8>>,
+        key: Key,
+    ) -> Result<Response<Bytes>, Error> {
+        let mut config = dgram::Config::new();
+        config.set_max_retries(self.retries as u8);
```
Comment (bal-e) [inline src/commands/notify.rs:386]: Maybe the `retries` field should just be a `u8`?

File: src/commands/notify.rs [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+        let connection = dgram::Connection::with_config(env.dgram(socket), config);
+        let connection = tsig::Connection::new(key, connection);
```
Comment (bal-e) [inline src/commands/notify.rs:388]: You can combine this with the `without_tsig` method by using `Box<dyn>`.

**Review by bal-e — COMMENTED** [review]
Really well done, @tertsdiepraam!  This is some complex code but I like the way you've managed to get through it!

File: src/commands/notify.rs (hunk 7 above)
Comment (ximon18) [reply src/commands/notify.rs:124]: FYI other tools also use `-y`, e.g. `dig`.

File: src/commands/notify.rs [hunk 11]
```diff
@@ -0,0 +1,710 @@
...
+            // This is different from ldns's default of
+            // hmac-md5.sig-alg.reg.int but we don't support that algorithm.
+            algorithm = "hmac-sha512";
+        }
+
+        // With dig TSIG keys are also specified with -y,
+        // but our format is: <name:key[:algo]>
+        //      and dig's is: [hmac:]name:key
+        //
+        // When we detect an unknown tsig algorithm in algo,
```
Comment (ximon18) [inline src/commands/notify.rs:80]:
```suggestion
        // When we detect an unknown TSIG algorithm in algo,
```

File: src/commands/notify.rs (hunk 8 above)
Comment (tertsdiepraam) [reply src/commands/notify.rs:322]: There's not a lot to document, I think. I added something, but I'm not sure what you'd like to see. What should I write?

Comment (bal-e) [reply src/commands/notify.rs:322]: Even a one-line note is enough.  As I'm reading code, I like having a quick summary idea of what each high-level function does ... and I couldn't figure out one for this function, at least not at a glance.  Something simple like "Send a notify packet to a specific server"?

**Review by bal-e — APPROVED** [review]
Nicely done, @tertsdiepraam!

File: src/commands/notify.rs [hunk 12]
```diff
@@ -142,20 +109,111 @@ pub struct Notify {
...
     servers: Vec<String>,
 }
 
+const LDNS_HELP: &str = "\
+usage: ldns-notify [other options] -z zone <servers>
+Ldns notify utility
+
+ Supported options:
+        -z zone         The zone
+        -I <address>    source address to query from
```
Comment (mozzieongit) [inline src/commands/notify.rs:125]: Maybe we should mention that this is currently unsupported?

File: src/commands/notify.rs [hunk 13]
```diff
@@ -142,20 +109,111 @@ pub struct Notify {
...
+
+        let mut parser = lexopt::Parser::from_args(args);
+
+        while let Some(arg) = parser.next()? {
+            match arg {
+                Arg::Short('z') => {
+                    let val = parser.value()?;
+                    zone = Some(parse_os("zone (-z)", &val)?);
+                }
+                Arg::Short('I') => todo!(),
```
Comment (mozzieongit) [inline src/commands/notify.rs:160]: Maybe this should be an error instead saying it's currently unsupported?

**Review by mozzieongit — CHANGES_REQUESTED** [review]
Last complaints, I swear! _... maybe_ :D

File: src/commands/notify.rs (hunk 13 above)
Comment (tertsdiepraam) [reply src/commands/notify.rs]: You're absolutely right of course!

File: src/commands/notify.rs (hunk 12 above)
Comment (tertsdiepraam) [reply src/commands/notify.rs]: Good idea!

**Comment by tertsdiepraam** [comment]
@mozzieongit do your worst! Give me that great feedback! 😄
